### PR TITLE
Add "toggle all with alt-click" to `batch-mark-files-as-viewed`

### LINKED ISSUE #3043: Add "toggle all with alt-click" to `batch-mark-files-as-viewed`
Follows #3007 

- [ ] Add alt-click to toggle all ("all _loaded_ files")
- [ ] Add a "in progress" indicator next to the counter, because the marking might take a few seconds:
	counter

### PR DESCRIPTION
**Pull request by yakov116**
1. LINKED ISSUES:
   Closes #3043

2. TEST URLS:
   https://github.com/sindresorhus/refined-github/pull/4044

3. SCREENSHOT:
Video_2021-03-12_103553

### CODE REVIEW

**Comment by yakov116** [comment]
@fregante did understand the issue correctly? You want that when clicking the progress bar it should mark them all as viewed?

**Comment by fregante** [comment]
It should toggle them. If you alt-click a checked item, it should uncheck it and uncheck all of them. 

If you alt-click an unchecked item, all of them should be checked.

**Comment by yakov116** [comment]
> If you alt-click a checked item, it should uncheck it and uncheck all of them.

If you alt check ANY item it should uncheck/check the ALL? So nothing to do with the progress bar?

**Comment by fregante** [comment]
The progress bar is unrelated to the logic. It’s only there because it might take a few seconds. Just refer to how `toggle-all-with-alt-click` works. It’s the same identical logic.

**Comment by yakov116** [comment]
...attempts to use new feature but its not correct :) the name is `toggle-everything-with-alt`

File: source/features/batch-mark-files-as-viewed.tsx [hunk 1]
```diff
@@ -87,11 +39,15 @@ function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
...
+function markAsViewedSelector(): string {
+	return '.js-reviewed-checkbox';
+}
+
 function init(): void {
 	// `mousedown` required to avoid mouse selection on shift-click
 	delegate(document, '.js-toggle-user-reviewed-file-form', 'mousedown', batchToggle);
 	delegate(document, '.js-toggle-user-reviewed-file-form', 'submit', remember);
-	delegate(document, '.diffbar-item progress-bar', 'click', markAll);
+	delegate(document, '.js-reviewed-checkbox', 'click', clickAll(markAsViewedSelector, true));
```
Comment (fregante) [inline source/features/batch-mark-files-as-viewed.tsx:50]: Neat!

File: source/features/toggle-everything-with-alt.tsx [hunk 2]
```diff
@@ -1,32 +1,56 @@
...
+// eslint-disable-next-line import/prefer-default-export
+export const clickAll = mem((selectorGetter: ((clickedItem: HTMLElement) => string), displayProgress?: boolean): EventHandler => {
 	return event => {
 		if (event.altKey && event.isTrusted) {
 			const clickedItem = event.delegateTarget;
 
 			// `parentElement` is the anchor because `clickedItem` might be hidden/replaced after the click
 			const resetScroll = preserveScroll(clickedItem.parentElement!);
-			clickAllExcept(selectorGetter(clickedItem), clickedItem);
+			clickAllExcept(selectorGetter(clickedItem), clickedItem, displayProgress);
```
Comment (fregante) [inline source/features/toggle-everything-with-alt.tsx:21]: But the progress is unrelated to this operation. I suggest 2 things:

- Have a self-contained toaster/spinner component, like: `const toast = new Toast()` and `toast.done()` to be called later, something like that. This is something that is likely to be reused in other batch features as well.
- Have the event handler in `batch-mark-files-as-viewed` call `new Toast(); clickAllBlahBlah; toast.done();`

**Review by fregante — CHANGES_REQUESTED** [review]

File: source/features/toggle-everything-with-alt.tsx (hunk 2 above)
Comment (yakov116) [reply source/features/toggle-everything-with-alt.tsx:21]: > * Have the event handler in [`batch-mark-files-as-viewed`](/sindresorhus/refined-github/blob/main/source/features/batch-mark-files-as-viewed.tsx) call `new Toast(); clickAllBlahBlah; toast.done();`

Going to give this only one attempt. (_all help wanted_) 🚑 

How do I make an event handler pass back to the main function (ClickedItem calls clickExcept) and then have it available to pass back? Do we have any feature examples I can look at?

Comment (fregante) [reply source/features/toggle-everything-with-alt.tsx:21]: I think it looks something like:

```ts
const markAsViewedSelector = clickAll((): string => '.js-reviewed-checkbox', true)

function handlerblah () {
	showToast
	markAsViewedSelector()
	hideToast
}

Comment (yakov116) [reply source/features/toggle-everything-with-alt.tsx]: 🤷 

I tried it but it does not want to mark them all if I do that

Comment (fregante) [reply source/features/toggle-everything-with-alt.tsx]: The current code is not acceptable though. Most clickAll operations take milliseconds to run and don’t need the toast. The toast does not belong in clickAll.

Comment (yakov116) [reply source/features/toggle-everything-with-alt.tsx]: Done

Comment (yakov116) [reply source/features/toggle-everything-with-alt.tsx]: https://github.com/sindresorhus/refined-github/pull/4044/files takes almost 3 seconds and freezes that page 😞

Comment (fregante) [reply source/features/toggle-everything-with-alt.tsx]: > [#4044 (files)](https://github.com/sindresorhus/refined-github/pull/4044/files) takes almost 3 seconds and freezes that page 😞

That's why the the toast is needed for toggle-all

Comment (yakov116) [reply source/features/toggle-everything-with-alt.tsx]: I tried I dont know. Currently the feture does not work and I cant figure out WHY?!

**Comment by fregante** [comment]
> 2\. TEST URLS:
>     This PR

This PR has 5 files, not a good test PR for a toast. I suggest https://github.com/sindresorhus/refined-github/pull/4044

Fixed:

gif

File: source/features/table-input.tsx [hunk 3]
```diff
@@ -48,9 +49,7 @@ function init(): void {
 						className="tooltipped tooltipped-n"
 						aria-label="Add a table"
 					>
-						<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg" className="octicon">
-							<path d="m2.75 1c-0.9665 0-1.75 0.7835-1.75 1.75v10.5c0 0.9665 0.7835 1.75 1.75 1.75h10.5c0.9665 0 1.75-0.7835 1.75-1.75v-10.5c0-0.9665-0.7835-1.75-1.75-1.75h-10.5zm0 1.5h4.5v4.75h-4.75v-4.5c0-0.13807 0.11193-0.25 0.25-0.25zm6 0h4.5c0.1381 0 0.25 0.11193 0.25 0.25v4.5h-4.75v-4.75zm-6.25 6.25h4.75v4.75h-4.5c-0.13807 0-0.25-0.1119-0.25-0.25v-4.5zm6.25 0h4.75v4.5c0 0.1381-0.1119 0.25-0.25 0.25h-4.5v-4.75z"/>
-						</svg>
+						<TableIcon/>
```
Comment (fregante) [inline source/features/table-input.tsx:52]: Unrelated

File: source/features/batch-mark-files-as-viewed.tsx [hunk 4]
```diff
@@ -38,10 +41,27 @@ function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
...
+}
+
+const markAsViewed = clickAll(markAsViewedSelector);
+
+async function onAltClick(event: delegate.Event<MouseEvent, HTMLElement>): Promise<void> {
+	if (event.altKey && event.isTrusted) {
+		new Toast().show();
+		await delay(30); // Without this, the Toast doesn't appear in time
+		markAsViewed(event);
+		new Toast().done();
```
Comment (fregante) [inline source/features/batch-mark-files-as-viewed.tsx:56]: It's not a new toast every time. It should be something like
```suggestion
		const toast = new Toast();
		await delay(30); // Without this, the Toast doesn't appear in time
		markAsViewed(event);
		toast.off();
```

File: source/github-helpers/toast.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	done(textContent = 'Bulk action processing complete.'): void {
+		this.destroy();
+		document.body.append(this.baseElement('Toast--success', <CheckIcon/>, textContent));
+		setTimeout(() => {
+			this.destroy();
+		}, 3000);
+	}
+
+	destroy(): void {
+		select('.rgh-toast')?.remove();
```
Comment (fregante) [inline source/github-helpers/toast.tsx:37]: It can't just select any toast, it should remove itself or else it could end up removing unrelated toasts of operations that are still in progress.

Comment (fregante) [reply source/github-helpers/toast.tsx:37]: Also this just removes the toast without animation, does GH do the same?

File: source/github-helpers/toast.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+		document.body.append(this.baseElement('Toast--loading', <ToastSpinner/>, textContent));
+	}
+
+	done(textContent = 'Bulk action processing complete.'): void {
+		this.destroy();
+		document.body.append(this.baseElement('Toast--success', <CheckIcon/>, textContent));
+		setTimeout(() => {
+			this.destroy();
+		}, 3000);
+	}
```
Comment (fregante) [inline source/github-helpers/toast.tsx:34]: Ohh this would look much better with #4058

File: source/helpers/icons.tsx [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import React from 'dom-chef';
+
+export function TableIcon(): JSX.Element {
```
Comment (fregante) [inline source/helpers/icons.tsx:3]: One icon per file

File: source/helpers/icons.tsx [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+	return (
+		<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg" className="octicon">
+			<path d="m2.75 1c-0.9665 0-1.75 0.7835-1.75 1.75v10.5c0 0.9665 0.7835 1.75 1.75 1.75h10.5c0.9665 0 1.75-0.7835 1.75-1.75v-10.5c0-0.9665-0.7835-1.75-1.75-1.75h-10.5zm0 1.5h4.5v4.75h-4.75v-4.5c0-0.13807 0.11193-0.25 0.25-0.25zm6 0h4.5c0.1381 0 0.25 0.11193 0.25 0.25v4.5h-4.75v-4.75zm-6.25 6.25h4.75v4.75h-4.5c-0.13807 0-0.25-0.1119-0.25-0.25v-4.5zm6.25 0h4.75v4.5c0 0.1381-0.1119 0.25-0.25 0.25h-4.5v-4.75z"/>
+		</svg>
+	);
+}
+
+export function ToastSpinner(): JSX.Element {
```
Comment (fregante) [inline source/helpers/icons.tsx:11]: I'd probably just inline it at this point. The name even suggests that it belongs to the toast

**Review by fregante — CHANGES_REQUESTED** [review]

File: source/helpers/icons.tsx (hunk 8 above)
Comment (fregante) [reply source/helpers/icons.tsx:11]: Or keep it in the toast’s file as it own function still

**Comment by yakov116** [comment]
> The PR is "ready for review" but no screenshot 🥺

I did not even think you were going to keep the PR open. Will work on this tomorrow.

File: source/github-helpers/toast.tsx (hunk 5 above)
Comment (yakov116) [reply source/github-helpers/toast.tsx:44]: > Also this just removes the toast without animation, does GH do the same?

They don't

File: source/features/batch-mark-files-as-viewed.tsx [hunk 9]
```diff
@@ -38,10 +42,30 @@ function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
...
+const markAsViewed = clickAll(markAsViewedSelector);
+
+async function onAltClick(event: delegate.Event<MouseEvent, HTMLElement>): Promise<void> {
+	if (event.altKey && event.isTrusted) {
+		const lastCheckboxCompleted = oneMutation(select.last(markAsViewedSelector(event.target as HTMLElement))!, {attributes: true});
+		const toast = new Toast();
+		toast.show();
+		await delay(30); // Without this, the Toast doesn't appear in time
+		markAsViewed(event);
+		await lastCheckboxCompleted; // Without this, done will be called too early
```
Comment (fregante) [inline source/features/batch-mark-files-as-viewed.tsx:56]: How early? I just checked #4044 on Chrome and I don't see a difference?

File: source/github-helpers/toast.tsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+			style={{zIndex: 101}}
+			className="rgh-toast position-fixed bottom-0 right-0 ml-5 mb-5 anim-fade-in fast Toast Toast--loading"
+		>
+			{iconWrapper}
+			{messageWrapper}
+		</div>
+	);
+
+	document.body.append(toast);
+	await delay(30); // Without this, the Toast doesn't appear in time
```
Comment (fregante) [inline source/github-helpers/toast.tsx:31]: Could this be replaced with `await new Promise(requestAnimationFrame)`?

File: source/features/batch-mark-files-as-viewed.tsx [hunk 11]
```diff
@@ -38,10 +41,28 @@ function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
...
+	const checked = (target as HTMLInputElement).checked ? ':not([checked])' : '[checked]';
+	return '.js-reviewed-checkbox' + checked;
+}
+
+const markAsViewed = clickAll(markAsViewedSelector);
+
+async function onAltClick(event: delegate.Event<MouseEvent, HTMLElement>): Promise<void> {
+	if (event.altKey && event.isTrusted) {
+		const lastCheckboxCompleted = oneMutation(select.last(markAsViewedSelector(event.delegateTarget))!, {attributes: true});
+		const hideToast = await showToast();
```
Comment (fregante) [inline source/features/batch-mark-files-as-viewed.tsx:54]: We could customize the message, maybe:

```suggestion
		const message = event.delegateTarget.checked ?
			'Marking visible files as viewed' :
			'Marking visible files as unviewed';
		const hideToast = await showToast(message);
```

And also in `hideToast`

File: source/features/batch-mark-files-as-viewed.tsx (hunk 9 above)
Comment (yakov116) [reply source/features/batch-mark-files-as-viewed.tsx:56]: Without it the 'completed' toast is never shown. (I just tried can provide git if needed)

File: source/github-helpers/toast.tsx (hunk 10 above)
Comment (yakov116) [reply source/github-helpers/toast.tsx:31]: Nope

File: source/features/batch-mark-files-as-viewed.tsx (hunk 9 above)
Comment (fregante) [reply source/features/batch-mark-files-as-viewed.tsx]: At first sight it works and I do see the toast, but also I noticed that the 3s timer starts **before** the completed toast appears because the page is still doing _layout_ or something 🤷‍♂️ so if that operation takes longer than the 3s, the completed toast is removed before it's even painted. 

This should be taken care by the toast component itself, so I moved this logic into its file.

**Review by fregante — APPROVED** [review]
It appears to be ok

**Comment by yakov116** [comment]
Works well (great job @fregante 😄 )

What do you say to my working change?

File: source/features/batch-mark-files-as-viewed.tsx [hunk 12]
```diff
@@ -38,10 +40,33 @@ function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
...
+			'Marking visible files as unviewed',
+		doneMessage: 'Marking files completed'
+	});
+}
+
 function init(): void {
 	// `mousedown` required to avoid mouse selection on shift-click
 	delegate(document, '.js-toggle-user-reviewed-file-form', 'mousedown', batchToggle);
 	delegate(document, '.js-toggle-user-reviewed-file-form', 'submit', remember);
+	delegate(document, '.js-reviewed-checkbox', 'click', onAltClick);
```
Comment (FloEdelmann) [inline source/features/batch-mark-files-as-viewed.tsx:69]: @yakov116 Could this be extended to also allow alt-clicking on the label?

The selector would then be `.js-reviewed-toggle`, but in `onAltClick`, the checkbox needs to be selected first.
